docs/cpu_templates/cpu-templates.md

```diff
@@ -0,0 +1,190 @@
+Intel T2CL and AMD T2A templates together aim to provide instruction set feature
+parity between CPUs running them, so they can form a heterogeneous fleet
+exposing the same instruction sets to the application.
+
+The V1N1 template is designed to represent ARM Neoverse V1 as ARM Neoverse N1.
+
+### Configuring static CPU templates
+
+Configuration of a static CPU template is performed via `/machine-config` API endpoint:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/machine-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+           "vcpu_count": 2,
+           "mem_size_mib": 1024,
+           "cpu_template": "T2CL"
+  }'
+```
+
+## Custom CPU templates
+
+Users can create their own CPU templates by creating a json file containing
+modifiers for CPUID, MSRs or ARM registers.
+
+> **Note**
+Creating custom CPU templates requires expert knowledge of
+CPU architectures. An inappropriate configuration may lead to guest crashes or
+making guests vulnerable to security attacks. Custom CPU templates must be
+tested thoroughly before use in production.
+
+> **Note**
+Having MSRs or ARM registers in the custom CPU template does
+not affect access permissions that guests will have to those registers.
+The access control is handled by KVM and is not influenced by CPU templates.
+
+In order to assist with creation and usage of CPU templates, there exists
+a CPU template helper tool.
+TODO: add a link to the CPU template helper tool when it is available.
+
+### Configuring custom CPU templates
+
+Configuration of a custom CPU template is performed via `/cpu-config` API endpoint.
+
+An example of configuring a custom CPU template on x86_64:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/cpu-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+        "cpuid_modifiers": [
+          {
+            "leaf": "0x1",
+            "subleaf": "0x0",
+            "flags": 0,
+            "modifiers": [
+              {
+                "register": "eax",
+                "bitmap": "0bxxxx000000000011xx00011011110010"
+              }
+            ]
+          }
+        ],
+        "msr_modifiers": [
+          {
+            "addr": "0x10a",
+            "bitmap": "0b0000000000000000000000000000000000000000000000000000000000000000"
+          }
+        ]
+      }'
+```
+
+An example of configuring a custom CPU template on ARM:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/cpu-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+        "reg_modifiers": [
+          {
+            "addr": "0x603000000013c020",
+            "bitmap": "0bxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000xxxxxxxxxxxx0000xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

Comment:
I don't think in reading the doc it is documented at this point that `x` passes through the value. It might be worth noting this.